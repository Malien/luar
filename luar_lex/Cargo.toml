[package]
name = "luar_lex"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
thiserror = "1.0"
lazy_static = "1.4.0"
logos = "0.12"
regex = "1.4"
num = "0.4"
quickcheck = { version = "1.0", optional = true }

[dev-dependencies]
test_util = { path = "../test_util" }
quickcheck = "1.0"
quickcheck_macros = "1.0"
insta = "1.3"
indoc = "1.0"