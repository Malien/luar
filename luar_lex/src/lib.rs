#[cfg(test)]
#[macro_use(quickcheck)]
extern crate quickcheck_macros;

pub mod format;
mod ident;
mod number_literal;
mod string_literal;
mod token;

pub type DynTokens = Box<dyn Iterator<Item = Token>>;

pub trait ToTokenStream {
    type Tokens: IntoIterator<Item = Token>;

    fn to_tokens(self) -> Self::Tokens;
}

impl<T: IntoIterator<Item = Token>> ToTokenStream for T {
    type Tokens = Self;

    fn to_tokens(self) -> Self::Tokens {
        self
    }
}

pub use ident::Ident;
pub use number_literal::NumberLiteral;
pub use string_literal::StringLiteral;
pub use token::Token;
