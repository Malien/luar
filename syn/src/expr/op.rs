use std::iter;

use luar_lex::{fmt_tokens, ToTokenStream, Token};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum BinaryOperator {
    // Precedence level 0
    And,
    Or,
    // Precedence level 1
    Less,
    Greater,
    LessOrEquals,
    GreaterOrEquals,
    NotEquals,
    Equals,
    // Precedence level 2
    Concat,
    // Precedence level 3
    Plus,
    Minus,
    // Precedence level 4
    Mul,
    Div,
    // Precedence level 5
    Exp,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum UnaryOperator {
    Minus,
    Not,
}

impl ToTokenStream for BinaryOperator {
    type Tokens = iter::Once<Token>;
    fn to_tokens(self) -> Self::Tokens {
        iter::once(match self {
            BinaryOperator::And => Token::And,
            BinaryOperator::Or => Token::Or,
            BinaryOperator::Less => Token::Less,
            BinaryOperator::Greater => Token::Greater,
            BinaryOperator::LessOrEquals => Token::LessOrEquals,
            BinaryOperator::GreaterOrEquals => Token::GreaterOrEquals,
            BinaryOperator::NotEquals => Token::NotEquals,
            BinaryOperator::Equals => Token::Equals,
            BinaryOperator::Concat => Token::Concat,
            BinaryOperator::Plus => Token::Plus,
            BinaryOperator::Minus => Token::Minus,
            BinaryOperator::Mul => Token::Mul,
            BinaryOperator::Div => Token::Div,
            BinaryOperator::Exp => Token::Exp,
        })
    }
}

impl ToTokenStream for UnaryOperator {
    type Tokens = iter::Once<Token>;
    fn to_tokens(self) -> Self::Tokens {
        iter::once(match self {
            UnaryOperator::Minus => Token::Minus,
            UnaryOperator::Not => Token::Not,
        })
    }
}

fmt_tokens!(BinaryOperator);
fmt_tokens!(UnaryOperator);

#[cfg(feature = "quickcheck")]
use quickcheck::{Arbitrary, Gen};

#[cfg(feature = "quickcheck")]
impl Arbitrary for UnaryOperator {
    fn arbitrary(g: &mut Gen) -> Self {
        match u8::arbitrary(g) % 2 {
            0 => UnaryOperator::Minus,
            1 => UnaryOperator::Not,
            _ => unreachable!(),
        }
    }
}

#[cfg(feature = "quickcheck")]
impl Arbitrary for BinaryOperator {
    fn arbitrary(g: &mut Gen) -> Self {
        match u8::arbitrary(g) % 14 {
            0 => BinaryOperator::And,
            1 => BinaryOperator::Or,
            2 => BinaryOperator::Less,
            3 => BinaryOperator::Greater,
            4 => BinaryOperator::LessOrEquals,
            5 => BinaryOperator::GreaterOrEquals,
            6 => BinaryOperator::NotEquals,
            7 => BinaryOperator::Equals,
            8 => BinaryOperator::Concat,
            9 => BinaryOperator::Plus,
            10 => BinaryOperator::Minus,
            11 => BinaryOperator::Mul,
            12 => BinaryOperator::Div,
            13 => BinaryOperator::Exp,
            _ => unreachable!(),
        }
    }
}

#[cfg(test)]
mod test {
    use crate::unspanned_lua_token_parser;

    use indoc::indoc;
    use logos::Logos;
    use luar_lex::Token;

    #[test]
    fn operator_precedence_1() {
        /*
        (
            (
                (
                    (
                        (
                            not (
                                (1)^(2)
                            )
                        ) * (
                            - (3)
                        )
                    ) + (
                        (4) / (5)
                    )
                ) .. (
                    (6) - (7)
                )
            ) < (8)
        ) and (
            (9) > (10)
        ) or (
            (11) == (12)
        )
        */
        static TEXT: &str = indoc! {"
            not 1^2 * - 3 + 4 / 5 .. 6 - 7 < 8 and 9 > 10 or 11 == 12
        "};
        let tokens: Vec<_> = Token::lexer(TEXT).collect();
        let expression = unspanned_lua_token_parser::expression(tokens).unwrap();
        insta::assert_debug_snapshot!(expression);
    }

    #[test]
    fn operator_precedence_2() {
        /*
        (
            (1) <= (2)
        ) and (
            (
                (3) + (
                    (
                        not (4)
                    ) * (
                        (5) ^ (6)
                    )
                )
            ) >= (
                (7) - (
                    (
                        - (8)
                    ) / (9)
                )
            )
        ) or (
            (10) ~= (
                (11) .. (12)
            )
        )
        */
        static TEXT: &str = indoc! {"
            1 <= 2 and 3 + not 4 * 5 ^ 6 >= 7 - - 8 / 9 or 10 ~= 11 .. 12
        "};
        let tokens: Vec<_> = Token::lexer(TEXT).collect();
        let expression = unspanned_lua_token_parser::expression(tokens).unwrap();
        insta::assert_debug_snapshot!(expression);
    }

    #[test]
    fn operator_precedence_3() {
        /*
        (
            (
                (1) <= (
                    (
                        (2) and (3)
                    ) + (
                        not (
                            (
                                (4) * (5)
                            )
                        ) ^ (6)
                    )
                )
            ) >= (
                (
                    (7) - (
                        (
                            - (8)
                        ) / (
                            (
                                (9) or (10)
                            ) ~= (11)
                        )
                    )
                ) .. (12)
            )
        )
        */
        static TEXT: &str = indoc! {"
            1 <= ((2 and 3) + not (4 * 5)) ^ 6 >= 7 - - 8 / ((9 or 10) ~= 11) .. 12
        "};
        let tokens: Vec<_> = Token::lexer(TEXT).collect();
        let expression = unspanned_lua_token_parser::expression(tokens).unwrap();
        insta::assert_debug_snapshot!(expression);
    }
}
