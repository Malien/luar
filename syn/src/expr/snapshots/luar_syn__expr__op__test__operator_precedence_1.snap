---
source: syn/src/expr/op.rs
expression: expression
---
BinaryOperator {
    lhs: BinaryOperator {
        lhs: BinaryOperator {
            lhs: BinaryOperator {
                lhs: BinaryOperator {
                    lhs: BinaryOperator {
                        lhs: UnaryOperator {
                            op: Not,
                            exp: BinaryOperator {
                                lhs: Number(
                                    NumberLiteral(
                                        1.0,
                                    ),
                                ),
                                op: Exp,
                                rhs: Number(
                                    NumberLiteral(
                                        2.0,
                                    ),
                                ),
                            },
                        },
                        op: Mul,
                        rhs: UnaryOperator {
                            op: Minus,
                            exp: Number(
                                NumberLiteral(
                                    3.0,
                                ),
                            ),
                        },
                    },
                    op: Plus,
                    rhs: BinaryOperator {
                        lhs: Number(
                            NumberLiteral(
                                4.0,
                            ),
                        ),
                        op: Div,
                        rhs: Number(
                            NumberLiteral(
                                5.0,
                            ),
                        ),
                    },
                },
                op: Concat,
                rhs: BinaryOperator {
                    lhs: Number(
                        NumberLiteral(
                            6.0,
                        ),
                    ),
                    op: Minus,
                    rhs: Number(
                        NumberLiteral(
                            7.0,
                        ),
                    ),
                },
            },
            op: Less,
            rhs: Number(
                NumberLiteral(
                    8.0,
                ),
            ),
        },
        op: And,
        rhs: BinaryOperator {
            lhs: Number(
                NumberLiteral(
                    9.0,
                ),
            ),
            op: Greater,
            rhs: Number(
                NumberLiteral(
                    10.0,
                ),
            ),
        },
    },
    op: Or,
    rhs: BinaryOperator {
        lhs: Number(
            NumberLiteral(
                11.0,
            ),
        ),
        op: Equals,
        rhs: Number(
            NumberLiteral(
                12.0,
            ),
        ),
    },
}
